felix/bpf-gpl/nat.h

```diff
@@ -70,19 +86,30 @@ static CALI_BPF_INLINE int skb_nat_l4_csum_ipv4(struct cali_tc_ctx *ctx, size_t
 		}
 	}
 
+	__wsum csum = 0;
+
+	bool csum_update = false;
 
-	if (ip_src_from != ip_src_to) {
+	if (!ip_equal(ip_src_from, ip_src_to)) {
 		CALI_DEBUG("L4 checksum update src IP from %x to %x\n",
-				bpf_ntohl(ip_src_from), bpf_ntohl(ip_src_to));
-		ret = bpf_l4_csum_replace(skb, off, ip_src_from, ip_src_to, flags | BPF_F_PSEUDO_HDR | 4);
-		CALI_DEBUG("bpf_l4_csum_replace(IP): %d\n", ret);
+				debug_ip(ip_src_from), debug_ip(ip_src_to));
+
+		csum = bpf_csum_diff((__u32*)&ip_src_from, sizeof(ip_src_from), (__u32*)&ip_src_to, sizeof(ip_src_to), csum);
```

Comment:
Every time we touch the checksum update logic it seems to break.  Wonder if there's a good way to check this.  the docs are clear as mud.

Reply:
we use one of the described case, I added some comments on how we do it.